Propose new commands FLUSHSTATE / EVALSTATE

luisfdez opened the pull request:
This PR proposes a couple of new commands to be added to collectd related to the state of metrics.

More than a final implementation, I would like first to get your input on them in order to polish its implementation or design.

The rationale behind exposing this commands is the following:

* **FLUSHSTATE**: provide a way to reset live the state of a metric. When an alarm is raised in a node and a recovery task is executed, it can be handy to flush the state in order to get a fresh notification of the current state of the metric. It relies on the new initial _unknown_ state.

* **EVALSTATE**: following the same scenario, depending how the plugins are configured it could happen that the next evaluation of a metric won't happen soon. This command allows the user to force the evaluation of a metric based on the current data in the cache by calling the threshold plugin. ((now the threshold plugin is harcoded, but if you find this command interesting, I guess the target plugin to be called should be parametrized)).

Apart from the scenario presented above, having these commands help keeping the existing cache data to retrigger some operations instead of having to restart collectd.

The current implementation is a rehash of existing plugins, so tips and recommendations are more than welcome to improve the code.

kwiatrox commented on src/utils/cmds/flushstate.c:
```diff
@@ -0,0 +1,106 @@
+    return CMD_PARSE_ERROR;
+  }
+
+  identifier_copy = sstrdup(argv[0]);
```
You can define variable at first use
`char *identifier_copy = sstrdup(argv[0]);`

kwiatrox commented on src/utils/cmds/flushstate.c:
```diff
@@ -0,0 +1,106 @@
+
+  identifier_copy = sstrdup(argv[0]);
+
+  status = parse_identifier(argv[0], &ret_flushstate->identifier.host,
```
You can define variable at first use
`int status = parse_....`

kwiatrox commented on src/utils/cmds/evalstate.c:
```diff
@@ -0,0 +1,105 @@
+    return CMD_PARSE_ERROR;
+  }
+
+  identifier_copy = sstrdup(argv[0]);
```
You can define variable at first use
`char *identifier_copy = sstrdup(argv[0]);`

kwiatrox commented on src/utils/cmds/evalstate.c:
```diff
@@ -0,0 +1,105 @@
+
+  identifier_copy = sstrdup(argv[0]);
+
+  status = parse_identifier(argv[0], &ret_evalstate->identifier.host,
```
You can define variable at first use
`int status = parse_....`

kwiatrox commented on src/collectdctl.c:
```diff
@@ -228,6 +230,31 @@ static int getval(lcc_connection_t *c, int argc, char **argv) {
+static int evalstate(lcc_connection_t *c, int argc, char **argv) {
+  lcc_identifier_t ident;
+
+  int status;
```
You can define variable at first use.

kwiatrox commented on src/collectdctl.c:
```diff
@@ -337,6 +364,31 @@ static int flush(lcc_connection_t *c, int argc, char **argv) {
+static int flushstate(lcc_connection_t *c, int argc, char **argv) {
+  lcc_identifier_t ident;
+
+  int status;
```
You can define variable at first use.

kwiatrox commented on src/libcollectdclient/client.c:
```diff
@@ -822,6 +822,90 @@ int lcc_flush(lcc_connection_t *c, const char *plugin, /* {{{ */
+
+  lcc_response_t res;
+
+  int status;
```
You can define variable at first use.

kwiatrox commented on src/libcollectdclient/client.c:
```diff
@@ -822,6 +822,90 @@ int lcc_flush(lcc_connection_t *c, const char *plugin, /* {{{ */
+
+  lcc_response_t res;
+
+  int status;
```
You can define variable at first use.

luisfdez commented:
> Thank you for your contribution.
> The code itself looks good in general, I've only added minor comments for code style.

Thanks @kwiatrox for taking a look to the code and I apology for the time it took me to look at it again.

I have just updated the code and added a destroy step to the commands.

I see the redhat test is failing but as per a comment from @mrunge in another PR it seems to be a problem with the CI that needs fixing.
